[IMP] dashboard: add full screen button to pivot

hokolomopo opened the pull request:
## Description:

description of this task, what is implemented and why it is implemented that way.

Task: [4497373](https://www.odoo.com/odoo/2328/tasks/4497373)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

LucasLefevre committed:
[IMP] dashboard: display popover outside the grid

Steps to reproduce:
- create a dashboard with a small amount of columns
  (less the the width of your screen, just like standard dashboard)
- on the right-most column, insert a formula with an error (=0/0)
- in dashboard mode, hover the error cell

=> the error message popover is display on the left of the cell, even though
there's room on the right.

It's not critical for those error messages, but a following commit adds a
popover widget to make a pivot full screen.

Task: 4543812

LucasLefevre committed:
[IMP] grid: open popover sooner

Currently, when hovering a cell with an error, a link, or pivot sorting
tooltip, the user have to wait for 300ms before the popover is displayed.
It feels sluggish (Gsheet is much faster) but most importantly, it makes
it hard to discover the full screen pivot feature because there's no hint
that the cell is hoverable and it's not natural to stop on a cell
for a full 300ms for no apparent reason.

Task: 4543812

hokolomopo committed:
[REF] spreadsheet: split `Spreadsheet` component

This commit extract the logic that displays a spreadsheet into a new
`SpreadsheetEditor` component, only keeping the main spreadsheet/dashboard
and `env` logic in the `Spreadsheet` component.

Task: 4497373

hokolomopo committed:
[IMP] model: disallow transport service in dashboard mode

With the next commits, we introduce a new feature that opens a
playground sheet in dashboard mode. The changes to this playground
should be discarded as soon as the user leaves the playground.

That feature is not compatible with a collaborative environment. The
changes in the playground should never be broadcast to other users,
and it's not possible in a collaborative environment to discard
changes in the playground when the user leaves it, since the current
session implementation only allows to undo one change at a time.

As features, the dashboard mode and collaborative didn't make sense
together anyway.

With this commit, we completely prevent the model to be in dashboard mode
with a transport service. That was already our principal use case in Odoo.

Task: 4497373

hokolomopo committed:
[FIX] model: finalize is called twice on undo/redo

When receiving a `revision-undone/redone` message, the model would
manually trigger the finalize. But this is wrong if the model is
currently dispatching a command, as the finalize should then be called
only when the command is done.

It happened if the model used the default `LocalTransportService`,
which is synchronous and thus handled the `revision-undone/redone`
message immediately, while is was still dispatching the
`REQUEST_UNDO/REDO` command.

Task: 4497373

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/o-spreadsheet/pull/6807.png)](https://mergebot.odoo.com/odoo/o-spreadsheet/pull/6807)

hokolomopo committed:
[IMP] cell popover: add additional functionalities

With this commit, the cell popovers can:
- Be animated when they appear.
- Be positioned on another cell than the hovered one.

Task: 4497373

hokolomopo committed:
[IMP] dashboard: add possibility to show full screen pivot

This commit introduces a new feature that allows users to view pivots
in full screen mode, creating a new playground sheet with the pivot
that the user can edit at will. All the changes will be discarded
when the user exits the playground to go back to the dashboard.

Task: 4497373

hokolomopo commented on src/components/full_screen_sheet/full_screen_sheet_store.ts:
```diff
@@ -0,0 +1,109 @@
+  private makePivotFullScreen(pivotId: UID) {
+    const sheetId = new UuidGenerator().smallUuid();
+
+    this.model.dispatch("CREATE_SHEET", {
+      position: this.getters.getSheetIds().length,
+      name: this.getters.getNextSheetName(),
```
So ... we want to discard all the changes done after in the playground. This is probably the less worse option we found, and it's not great (changes in model/sheetview were necessary to make multiple undo in a single command work when we don't have a transport service).

But the alternatives are somehow worse:
- creating  new model to display the playground. We'd need a way to not reload every pivot (using value from the other model), because that can take a while in dahsboard
- similar to the first version of version history, do a `session.revision.goTo(revisionId)`. It doesn't really bring much over spamming UNDO, and we'd need to remember the changed datasources and refresh them by hand. (the plugins handle that in UNDO)

pro-odoo replied:
I'm still unsure of the chosen strategy, IMO the creation of a new model is a cleaner way to do:
- No changes in the model/sheetview
- No magic undo needed
- No restriction to have a transport service in dashboard (useless for now, but it could be useful if your dashboard is in an external screen and you do not want to refresh everything when you change something in the dashboard.)
- No need to make a screenshot of the dashboard
- No scroll position loosing
- Almost all your comments are outdated :D 

The only downside of this solution is the need to reload all the data. For that, we have multiple solutions:
- Make a cache of the read_group/search_read. I wonder if we could re-use the cache version of framework js...  In that case, it's probably easy. The only thing we have to make sure is to invalidate this cache when the commands `REFRESH_ALL_DATA_SOURCE`, `REFRESH_PIVOT` are triggered. 
- Do nothing => Ok bad solution, but still
- Find a way to share PivotModel and ListDataSource to the newly created model

What do you think ?

hokolomopo commented on src/components/helpers/dom_helpers.ts:
```diff
@@ -259,3 +259,50 @@ function isOtherMobileOS() {
   return isAndroid() || isIOS() || isOtherMobileOS();
 }
+
+/** Return a base64 representation of the current spreadsheet, with the grid canvas and all the figures */
+export function getSpreadsheetAsBase64() {
```
So ... when we open a pivot in fullscreen we use a dialog-like style with a backdrop. But we cannot have two separate viewports, so we convert the spreadsheet to a png, and use that as background image 🙃

hokolomopo commented on src/components/full_screen_sheet/full_screen_sheet.ts:
```diff
@@ -0,0 +1,46 @@
+    });
+  }
+
+  exitFullScreen() {
+    this.fullScreenSheetStore.exitFullScreen();
+    // ADRM TODO DISCUSS WITH RAR: Manually resize the sheet here avoid having a flicker when exiting the full screen, because the
+    // first render has the wrong size, then the resize observer is triggered and the size is fixed for the next render.
+    // Kinda ugly. Not sure if we want to keep it. Not sure what alternative we have.
+    const { width, height } = getRefBoundingRect(this.spreadsheetRef);
+    this.env.model.dispatch("RESIZE_SHEETVIEW", { width, height, gridOffsetX: 0, gridOffsetY: 0 });
+  }
```
🙈

hokolomopo commented on src/model.ts:
```diff
@@ -690,9 +701,16 @@ export class Model extends EventBus<any> implements CommandDispatcher {
+      );
+    }
     // @ts-ignore For testing purposes only
     this.config.mode = mode;
-    this.trigger("update");
+    if (this.status === Status.Ready) {
+      this.trigger("update");
+    }
```
Probably need to change so we can change the mode of the model in a clean way.

And aren't all those `if (this.status === Status.Ready) {` pretty and cool ? 🙈 🙉

hokolomopo commented on src/components/full_screen_sheet/full_screen_sheet.scss:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @keyframes growIn {
+      /* Hack to have the first keyFrame be the final size of the content, because resizeObservers won't react to animation afterwards */
+      0% {
+        transform: scale(1);
+        visibility: hidden;
+      }
+      1% {
+        transform: scale(0.8);
+        visibility: visible;
+      }
```
yeah ... 🙈

hokolomopo replied:
<img width="960" height="1440" alt="Image" src="https://github.com/user-attachments/assets/1685f092-21c5-4bfd-9ec6-6a8fb66888d0" />

hokolomopo reviewed (commented):
So there's a bunch of weird/strange/magic code in there. But for most of that, I don't think we have much of a choice to reach the specs 😔

pro-odoo reviewed (changes requested):
Not confortable with the chosen solution, see my answer to your comment

rrahir commented on src/types/commands.ts:
```diff
@@ -1028,6 +1029,15 @@ export interface SplitPivotFormulaCommand extends PositionDependentCommand {
   pivotId: UID;
 }
 
+export interface MakePivotFullScreen {
```
```suggestion
export interface MakePivotFullScreenCommand {
```

rrahir commented on src/components/full_screen_sheet/full_screen_sheet_store.ts:
```diff
@@ -0,0 +1,123 @@
+    // AUTORESIZE in separate command because we need to wait for a finalize() to have an evaluation
+    if (this.fullScreenSheetId) {
+      const cols = this.getters.getPivot(pivotId).getExpandedTableStructure().numberOfColumns;
+      this.model.dispatch("AUTORESIZE_COLUMNS", {
```
:puke:
